src/storage/src/render/upsert.rs

```diff
@@ -765,52 +795,104 @@ where
+                        // Data within this set events strictly beyond this time are staged as
+                        // normal.
+                        //
+                        // This is a load-bearing optimization, as it is required to avoid buffering
+                        // the entire source snapshot in the `stash`.
+                        if prevent_snapshot_buffering
+                            && output_cap.time().cmp(event_time) == Ordering::Equal
+                        {
+                            partial_drain_time = Some(event_time.clone());
+                        }
                     }
+                    AsyncEvent::Progress(upper) => {
+                        tracing::trace!(?upper, "received progress in upsert");
+                        // Ignore progress updates before the `resume_upper`, which is our initial
+                        // capability post-snapshotting.
+                        if PartialOrder::less_than(&upper, &resume_upper) {
+                            continue;
+                        }
 
-                    drain_staged_input::<_, G, _, _, _>(
-                        &mut stash,
-                        &mut commands_state,
-                        &mut output_updates,
-                        &mut multi_get_scratch,
-                        &upper,
-                        &mut error_emitter,
-                        &mut state,
-                    )
-                    .await;
-
-                    // Emit the _consolidated_ changes to the output.
-                    output_handle
-                        .give_container(&output_cap, &mut output_updates)
+                        // Disable the partial drain as this progress event covers
+                        // the `output_cap` time.
+                        partial_drain_time = None;
+                        drain_staged_input::<_, G, _, _, _>(
+                            &mut stash,
+                            &mut commands_state,
+                            &mut output_updates,
+                            &mut multi_get_scratch,
+                            DrainStyle::ToUpper(&upper),
+                            &mut error_emitter,
+                            &mut state,
+                        )
                         .await;
-                    if let Some(ts) = upper.as_option() {
-                        output_cap.downgrade(ts);
+
+                        // Emit the _consolidated_ changes to the output.
+                        output_handle
+                            .give_container(&output_cap, &mut output_updates)
+                            .await;
+                        if let Some(ts) = upper.as_option() {
+                            output_cap.downgrade(ts);
+                        }
+                        input_upper = upper;
                     }
-                    input_upper = upper;
                 }
+                let events_processed = i + 1;
```

Comment:
What is this `+1` for?

Reply:
`i` is 0-indexed cause its from `enumerate()`